File: engine/normalize.py
"""Deterministic field normalisation. No LLM."""
import re

LEGAL_SUFFIXES = (
    r"\b(sdn\.?\s*bhd\.?|pte\.?\s*ltd\.?|co\.?,?\s*ltd\.?|limited|"
    r"llc|fz-?llc|fze|gmbh|inc\.?|corp\.?|corporation|ltd\.?|bhd\.?|sdn\.?)\b"
)

EXACT_BLANKS = {"???", "tba", "tbd", "n/a", "na", "nil", "unknown", ""}


def collapse_ws(value):
    return re.sub(r"\s+", " ", (value or "").replace("\xa0", " ")).strip()


def is_blank(value):
    v = collapse_ws(value).lower().strip(".:- ")
    if not v:
        return True
    compact = v.replace(" ", "")
    if compact in {"???", "______", "_____", "____", "___"}:
        return True
    if compact and set(compact) <= set("_?.-"):
        return True
    return v in EXACT_BLANKS


def company_key(value):
    v = collapse_ws(value).upper()
    v = re.sub(LEGAL_SUFFIXES, "", v, flags=re.I)
    v = re.sub(r"[^A-Z0-9 ]+", " ", v)
    return collapse_ws(v)


def port_code(value):
    v = collapse_ws(value).upper()
    codes = re.findall(r"\(([A-Z]{3,5})\)", v)
    return codes[-1] if codes else None


def port_key(value):
    """City/place name, ignoring a trailing UN/LOCODE in parentheses.

    The dataset sometimes plants a different city while keeping the same
    code, so comparing codes alone misses real mismatches.
    """
    v = collapse_ws(value).upper()
    v = re.sub(r"\(([A-Z]{3,5})\)\s*$", "", v).strip()
    first = v.split(",")[0]
    first = re.sub(r"\([^)]*\)", " ", first)
    return collapse_ws(re.sub(r"[^A-Z0-9 ]+", " ", first))


def ports_match(a, b):
    if is_blank(a) or is_blank(b):
        return False
    ka, kb = port_key(a), port_key(b)
    if ka and kb:
        if ka == kb or ka in kb or kb in ka:
            return True
        return False
    ca, cb = port_code(a), port_code(b)
    return bool(ca) and ca == cb


def container_count(value):
    v = collapse_ws(value)
    if is_blank(v):
        return None
    m = re.search(r"(\d+)\s*[xX]", v)
    if m:
        return int(m.group(1))
    m = re.search(r"\d+", v.replace(",", ""))
    return int(m.group(0)) if m else None


def weight_kg(value):
    v = collapse_ws(value)
    if is_blank(v):
        return None
    lower = v.lower()
    m = re.search(r"([\d,.]+)", v.replace(" ", ""))
    if not m:
        m = re.search(r"([\d,.]+)", v)
    if not m:
        return None
    num = float(m.group(1).replace(",", ""))
    if "lb" in lower or "pound" in lower:
        num *= 0.453592
    elif re.search(r"\bmt\b|metric ton|tonne", lower):
        num *= 1000
    return round(num, 3)


def names_match(a, b):
    if is_blank(a) or is_blank(b):
        return False
    ka, kb = company_key(a), company_key(b)
    if not ka or not kb:
        return False
    if ka == kb:
        return True
    leftover = None
    if ka in kb:
        leftover = collapse_ws(kb.replace(ka, " "))
    elif kb in ka:
        leftover = collapse_ws(ka.replace(kb, " "))
    else:
        return False
    # Distinct legal entities that share a prefix (e.g. APRIL ... vs APRIL ... MIDDLE EAST)
    if leftover in {"MIDDLE EAST", "FAR EAST"} or leftover.startswith("MIDDLE EAST"):
        return False
    return True

File: web/lib/normalize.js
/** Same as engine/normalize.py. */

const LEGAL_SUFFIXES =
  /\b(sdn\.?\s*bhd\.?|pte\.?\s*ltd\.?|co\.?,?\s*ltd\.?|limited|llc|fz-?llc|fze|gmbh|inc\.?|corp\.?|corporation|ltd\.?|bhd\.?|sdn\.?)\b/gi;

const EXACT_BLANKS = new Set(["???", "tba", "tbd", "n/a", "na", "nil", "unknown", ""]);

export function collapseWs(value) {
  return (value || "").replace(/\u00a0/g, " ").replace(/\s+/g, " ").trim();
}

export function isBlank(value) {
  const v = collapseWs(value).toLowerCase().replace(/^[.:\-\s]+|[.:\-\s]+$/g, "");
  if (!v) return true;
  const compact = v.replace(/ /g, "");
  if (["???", "______", "_____", "____", "___"].includes(compact)) return true;
  if (compact && /^[_\?.\-]+$/.test(compact)) return true;
  return EXACT_BLANKS.has(v);
}

export function companyKey(value) {
  let v = collapseWs(value).toUpperCase();
  v = v.replace(LEGAL_SUFFIXES, "");
  v = v.replace(/[^A-Z0-9 ]+/g, " ");
  return collapseWs(v);
}

export function portCode(value) {
  const v = collapseWs(value).toUpperCase();
  const codes = [...v.matchAll(/\(([A-Z]{3,5})\)/g)].map((m) => m[1]);
  return codes.length ? codes[codes.length - 1] : null;
}

export function portKey(value) {
  let v = collapseWs(value).toUpperCase();
  v = v.replace(/\(([A-Z]{3,5})\)\s*$/, "").trim();
  let first = v.split(",")[0];
  first = first.replace(/\([^)]*\)/g, " ");
  return collapseWs(first.replace(/[^A-Z0-9 ]+/g, " "));
}

export function portsMatch(a, b) {
  if (isBlank(a) || isBlank(b)) return false;
  const ka = portKey(a);
  const kb = portKey(b);
  if (ka && kb) {
    if (ka === kb || ka.includes(kb) || kb.includes(ka)) return true;
    return false;
  }
  const ca = portCode(a);
  const cb = portCode(b);
  return Boolean(ca) && ca === cb;
}

export function containerCount(value) {
  const v = collapseWs(value);
  if (isBlank(v)) return null;
  const x = v.match(/(\d+)\s*[xX]/);
  if (x) return parseInt(x[1], 10);
  const n = v.replace(/,/g, "").match(/\d+/);
  return n ? parseInt(n[0], 10) : null;
}

export function weightKg(value) {
  const v = collapseWs(value);
  if (isBlank(v)) return null;
  const lower = v.toLowerCase();
  let m = v.replace(/ /g, "").match(/([\d,.]+)/);
  if (!m) m = v.match(/([\d,.]+)/);
  if (!m) return null;
  let num = parseFloat(m[1].replace(/,/g, ""));
  if (lower.includes("lb") || lower.includes("pound")) num *= 0.453592;
  else if (/\bmt\b|metric ton|tonne/.test(lower)) num *= 1000;
  return Math.round(num * 1000) / 1000;
}

export function namesMatch(a, b) {
  if (isBlank(a) || isBlank(b)) return false;
  const ka = companyKey(a);
  const kb = companyKey(b);
  if (!ka || !kb) return false;
  if (ka === kb) return true;
  let leftover = null;
  if (kb.includes(ka)) leftover = collapseWs(kb.replace(ka, " "));
  else if (ka.includes(kb)) leftover = collapseWs(ka.replace(kb, " "));
  else return false;
  if (leftover === "MIDDLE EAST" || leftover === "FAR EAST" || leftover.startsWith("MIDDLE EAST")) {
    return false;
  }
  return true;
}
